### ISSUE TITLE
Project introspection idea

### ISSUE THREAD

**Issue by Kludex** [alex-oleshkevich/starlette_boot#1]
I'm going to share the idea that I had, which I didn't implement yet.

I want to have a very extensible tool, and I don't want to rely on static layouts. People may disagree with small things, and change their project structure, which would cause "naive" tools to fail. What I thought is to identify what I call "entities" in a project, and each entity has a metadata - and new custom entities can be created.

Let's go to an example. Assume we have the following project layout:

```bash
bootstrap/
├── app/
│   ├── __init.py__
│   ├── main.py
│   ├── config.py
│   └── database.py
└── Dockerfile
```
We would infer the entities on the project, and save it in a cache locally. Some entities I can see:

Entity: **Application**
Metadata: location (`main.py`), package (FastAPI/Starlette), middlewares

Entity: **Configuration**
Metadata: location (`config.py`), `kind` (`BaseSettings`, `Starlette.config`)

Entity: **Database**
Metadata: location `database.py`, `kind` (PostgreSQL, SQLite), `package` (SQLAlchemy, other)

---

Metadata can be more precise to make it easier to add plugins around it.

Then, if we have a Sentry plugin, it would interact with the Application entity, and include a middleware to the middleware metadata.

Another example, the alembic plugin, it would interact with the Database entity, and the entity that is in charge of startup command (?).

---

This is food for thoughts.

**Comment by alex-oleshkevich** [comment]
Presets. The tool itself can be generic and not opinionated. The presets it includes can vary from three file project to a full stack one.

Can you give more examples, may be in pseudo code about the introspection? I am not sure I completely understand how you see it working.

**Comment by Kludex** [comment]
I can try.

Assume the following application:
```python
# main.py

from fastapi import FastAPI

app = FastAPI()
```

We'd have to build the metadata about this project, using an AST/[CST](https://libcst.readthedocs.io/en/latest/) package, in a way that we can make it extensible for plugins to be attached.

<details>
 <summary>The above app has the following CST</summary>

```python
Module(
    body=[
        SimpleStatementLine(
            body=[
                ImportFrom(
                    module=Name(
                        value='fastapi',
                        lpar=[],
                        rpar=[],
                    ),
                    names=[
                        ImportAlias(
                            name=Name(
                                value='FastAPI',
                                lpar=[],
                                rpar=[],
                            ),
                            asname=None,
                            comma=MaybeSentinel.DEFAULT,
                        ),
                    ],
                    relative=[],
                    lpar=None,
                    rpar=None,
                    semicolon=MaybeSentinel.DEFAULT,
                    whitespace_after_from=SimpleWhitespace(
                        value=' ',
                    ),
                    whitespace_before_import=SimpleWhitespace(
                        value=' ',
                    ),
                    whitespace_after_import=SimpleWhitespace(
                        value=' ',
                    ),
                ),
            ],
            leading_lines=[],
            trailing_whitespace=TrailingWhitespace(
                whitespace=SimpleWhitespace(
                    value='',
                ),
                comment=None,
                newline=Newline(
                    value=None,
                ),
            ),
        ),
        SimpleStatementLine(
            body=[
                Assign(
                    targets=[
                        AssignTarget(
                            target=Name(
                                value='app',
                                lpar=[],
                                rpar=[],
                            ),
                            whitespace_before_equal=SimpleWhitespace(
                                value=' ',
                            ),
                            whitespace_after_equal=SimpleWhitespace(
                                value=' ',
                            ),
                        ),
                    ],
                    value=Call(
                        func=Name(
                            value='FastAPI',
                            lpar=[],
                            rpar=[],
                        ),
                        args=[],
                        lpar=[],
                        rpar=[],
                        whitespace_after_func=SimpleWhitespace(
                            value='',
                        ),
                        whitespace_before_args=SimpleWhitespace(
                            value='',
                        ),
                    ),
                    semicolon=MaybeSentinel.DEFAULT,
                ),
            ],
            leading_lines=[
                EmptyLine(
                    indent=True,
                    whitespace=SimpleWhitespace(
                        value='',
                    ),
                    comment=None,
                    newline=Newline(
                        value=None,
                    ),
                ),
            ],
            trailing_whitespace=TrailingWhitespace(
                whitespace=SimpleWhitespace(
                    value='',
                ),
                comment=None,
                newline=Newline(
                    value=None,
                ),
            ),
        ),
    ],
    header=[
        EmptyLine(
            indent=True,
            whitespace=SimpleWhitespace(
                value='',
            ),
            comment=None,
            newline=Newline(
                value=None,
            ),
        ),
    ],
    footer=[],
    encoding='utf-8',
    default_indent='    ',
    default_newline='\n',
    has_trailing_newline=True,
)
```

</details>

Let's say that we want to include `CORSMiddleware` via component plugin. For that, the `CORSMiddleware` plugin needs to be injected in the `FastAPI` constructor, and we need an easy way to build this.

The way I see it, to build the plugin, would look like this:

```python
from starlette_boot import Context

def cors_middleware_plugin(context: Context) -> None:
    entity: ApplicationEntity = context.get_entity("app")
    entity.add_middleware(CORSMiddleware, params={})
    # context can have methods like `create_file()`, `get_structure()`, `list_entities()`.
```

The `context.get_entity` would return an `Entity` object that would look like:
```python
class ApplicationEntity(Entity):
    path: Path  # main.py
    app_node: Call  # The CST object
    module_node: Module  # The CST object

    def add_middleware(middleware, params):
        # here we need to add the middleware to the `app_node`, and the import to the `module_node`
```

The `app_node` above would be like this:
```python
Call(
   func=Name(
      value='FastAPI',
      lpar=[],
      rpar=[],
   ),
   args=[],
   lpar=[],
   rpar=[],
   whitespace_after_func=SimpleWhitespace(
       value='',
   ),
   whitespace_before_args=SimpleWhitespace(
       value='',
   ),
)
```

Basically, the `entities` have an interface to allow them to be extended, and the `context` holds the information about the project, and both provide an interface to interact with them.

---

The external plugins can be found by the `starlette_boot` by being added as an [entrypoint](https://packaging.python.org/en/latest/specifications/entry-points/), and we'd also add native plugins (like all the ones you said on the email that you think are super important). Analogous to what [pytest does](https://docs.pytest.org/en/7.1.x/how-to/writing_plugins.html#making-your-plugin-installable-by-others).

**Comment by alex-oleshkevich** [comment]
The idea is cool and I see it like that too but I don't see how we can do this. For example adding the middleware. Fast API and Starlette do this differently. Moreover, using Starlette you have two ways to add the middleware - passing to constructor or via `middleware` variable. 
These issues are gone when:
1. Fast API and Starlette transformers are separate
2. There is a well know project layout so the tool knows where too look for the files.

I also played with the entities idea and failed to implement it. I found that providing a low level utilities is more flexible and extendable.
```python
context.apply_codemod('myproject.middleware', [AddMiddlewareCodemod('starlette.middleware.SessionMiddleware')])
```

However, this also fails quickly if your mw needs parameters or other preconditions.
So my working idea is that each component provides a codemod for own needs. 

Or do you have something in your mind?

**Comment by alex-oleshkevich** [comment]
Probably, we would make the tool as a very generic on without anything prebuilt. Instead, project scaffolds and component should be fully remote, much like cookiecutter does.

```bash
starlette new kludex/scaffold
starlette new alex.oleshkevich/scaffold 
```

Will make two projects but using different structure and deps.

**Comment by alex-oleshkevich** [comment]
The project preset is a list of components to install. And the component is a feature itself (scaffold, database add-on, etc).

**Comment by alex-oleshkevich** [comment]
The component can be like this

1. An optional template for cookiecutter
2. `compoment.py` module

The component module has Component class where we do keep generator logic.

Plus, for fastapi there could be another base class specific to it. Not compatible with Starlette.

In `Context` we can provide a list of project deps so you can infer which framework the project uses.
